File: Arrays/Easy/Pascals Triangle/Pascals_Triangle.java
import java.util.ArrayList;
import java.util.List;

public class Pascals_Triangle {

    static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> triangle = new ArrayList<List<Integer>>();
        if (numRows <= 0) {
            return triangle;
        }
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<Integer>();
            for (int j = 0; j < i + 1; j++) {
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    row.add(triangle.get(i - 1).get(j - 1) + triangle.get(i - 1).get(j));
                }
            }
            triangle.add(row);
        }
        return triangle;
    }

    public static void main(String[] args) {

        int numRows = 5;

        List<List<Integer>> pascalTriangle = new ArrayList<List<Integer>>();
        pascalTriangle = generate(numRows);

        for (int i = 0; i < pascalTriangle.size(); i++) {
            for (int j = 0; j < pascalTriangle.get(i).size(); j++) {
                System.out.print(pascalTriangle.get(i).get(j) + " ");
            }
            System.out.println();
        }

    }
}

File: 1. Arrays/Easy/Pascals Triangle/Pascals_Triangle.cpp
#include <bits/stdc++.h>
using namespace std;

/* ---------------------- Solution Class -------------------- */
class Solution
{
private:
public:
    /* Method to print Pascal's Triangle. */
    void print2DVec(vector<vector<int>> arr)
    {
        for (int i = 0; i < arr.size(); i++)
        {

            for (int j = 0; j < arr.size() - i - 1; j++)
            {
                cout << " ";
            }

            for (int j = 0; j < arr[i].size(); j++)
            {
                cout << arr[i][j] << " ";
            }
            cout << endl;
        }
        cout << endl;
    }

    /* Method to generate Pascal's Triangle. */
    // Time Complexity:- O(row * col).
    // Space Complexity:- O(row * col).
    vector<vector<int>> generate(int numRows)
    {
        // 2d vector to strore answer.
        vector<vector<int>> pascalTriangle(numRows);

        for (int i = 0; i < numRows; i++)
        {
            for (int j = 0; j <= i; j++)
            {
                if (j == 0 || j == i)
                {
                    pascalTriangle[i].push_back({1});
                }
                else
                {

                    pascalTriangle[i].push_back(pascalTriangle[i - 1][j - 1] + pascalTriangle[i - 1][j]);
                }
            }
        }

        return pascalTriangle;
    }
};

int main()
{
    Solution S;

    int numRows = 5;

    vector<vector<int>> pascalTriangle = S.generate(numRows);

    S.print2DVec(pascalTriangle);
}